capi-adoption/README.md

```diff
@@ -0,0 +1,114 @@
+
+We should have a certain level of defaulting for a set of versions we know to work well together, but we must not restrict customers from leveraging the flexibility of Cluster API.
+
+When we have already agreed to perform automatic minor and patch upgrades, we can equally just adopt the upstream workflow and update management cluster controllers in-place.
+
+## Proposal for alternative approach to provide Giant Swarm platform with upstream Cluster API
+
+To pick up the key points from current problems and to fully align with upstream in order to simplify our daily work and allow customers more flexibility while improving stability, here are proposed steps forward.
+
+### Let go of releases and multiple versions in same management cluster
+
+Currently the major source of friction with upstream and the source of complexity on our platform is coming from immutable releases and the need to run multiple versions of everything in the same Kubernetes cluster.
+
+Since the upstream Cluster API drastically changes the way how things are configured, we should leverage it, not work against it. Even though we have managed to contribute code upstream to support our use case, the community sees these contributions as workarounds to avoid blocking us. But we’d be always going in a different direction, increasing the effort and cost to fully implement CAPI and be aligned with the project.
+There are less risks for failure when different parts are managed separately and when active change of resources is always a direct consequence of active change on CR[s]. Management cluster operators can be upgraded when there are fixes, but they must not automatically change any of the reconciled resources. When there is a hot security fix for Kubernetes, it’s enough to update only the Kubernetes version. When workload cluster applications need updating, customers can perform it at their will via our App management infrastructure.
+
+Instead of strictly trying to keep control of release components, we should have sane defaults for CR templating. This would be our way to indicate “we have tested these components in this combination and found them to be rock solid, use them”.
+Focusing on templating with sane defaults rather than enforcing certain values would also mean getting rid of admission controllers that we currently use to validate and default CR content, simplifying the Kubernetes configuration.
+
+Allowing customers to customize those sane defaults would ease the testing of new combinations. For example, since upgrading Kubernetes to a new version would be only a matter of changing the required CRs, customers can test their workloads running on the new Kubernetes version on the cluster they decide at their own pace. And they would be testing that change in isolation. There would be no changes in the infrastructure resources or the installed apps.
+
+### Change our mindset when implementing operators
```

Comment:
I like this in general but I have some thoughts:
Currently we write opinionated infrastructure as code - meaning that we do changes to infrastructure through our operator code where the "opinion" is hardcoded in our operators.
Changing our mindset as described here would mean that we have more of these opinionated decisions in configuration through CRs.

I believe that CAPI providers also have a certain opinionated infra decisions in code. Other parts are then configurable through CRs. From what I see it leads to changes being bundled into minors - a lot of changes are being bundled into `v1alpha4` for example because the opinionated infra can not change without breaking changes in the CRDs.

At the same time too much configurability through the CR could make them hard to understand or hard to use.

To me this is about a certain balance between these approaches because either extreme has drawbacks in my mind.

Reply:
I would really switch _minor_ into _major_ in our terminologies here because that's what they really are in kubernetes ecosystem. I don't consider it as an issue that there are changes when major version jumps. Quite contrary, they are expected and therefore we need to behave accordingly. The patch updates then bring occasionally small improvements and most of the time fix bugs.

Regarding configurability, I only have good experience from CAPZ at the moment, but the opinionated infra decisions are quite small implementation details. In my experience the CAPZ controllers have been made to be very flexible and for most parts of the infrastructure you can actually "bring your own" into the setup. This allows users to create very different looking clusters on same installation, but at the same time, this also makes many of the silent modifications in infrastructure explicit.

We have quite a bit of experience from some users silently modifying the infrastructure of reconciled clusters and with the Cluster API / CAPZ approach many of these modifications can be performed via CR configuration and therefore made explicit. We could embrace this and build a tool which can show a clear diff between our defaulted infrastructure and currently configured one, in order to more easily understand what's going on when something doesn't work as expected.